# WP 2

## Introduction 

The main task of the WP2 is to annotate the clusters obtained from WP1 using the central repository for management and application of marker gene lists. The goal is to perform an appropiate computational analysis to extract meaningful biological insights of the single cell data.

## Task


## Contents
- [Backgrounds.md](archive/Backgrounds.md): The knowledge required to understand and use this project. 
- [Install.md](./Install.md): 
- [codes](./codes/):
  - [auto_annotation.py](./codes/auto_annotation.py):
  - [auto_annotation_notebook.ipynb](./codes/auto_annotation_notebook.ipynb):
  - [homology_notebook.ipynb](./codes/homology_notebook.ipynb):
  - [submit_new_marker_list_notebook.ipynb](./codes/submit_new_marker_list_notebook.ipynb):
- [data](./data/):
  - [DataSheet1.XLSX](./data/DataSheet1.XLSX)
  - [DataSheet2.XLSX](./data/DataSheet2.XLSX)
  - [2024_02_17](./data/2024_02_17/):

## Installation
To start the project, follow the instructions in [Install.md](./Install.md) to set up your development environment.

## References and Materials

The following is a list of scientific publications, technical notes, and a university thesis that were used as main sources of literature to research the most important concepts related to the project. The list also includes the paper from which the material with the zebrafish marker Excel files was taken.
We recommend reading the publications in case the person is not familiar with the subject.


1. Cell Type Annotation Strategies for Single Cell ATAC-Seq Data
   - Last modified March 30, 2022
   - https://www.10xgenomics.com/support/single-cell-atac/documentation/steps/sequencing/cell-type-annotation-strategies-for-single-cell-atac-seq-data
   - Explores three different strategies for annotating cell types in single-cell ATAC-seq data 

2. From reads to insight: a hitchhiker´s guide to ATAC-seq data analysis
   - Authors: Feng Yan, David R. Powell, David J. Curtis & Nicholas C. Wong
   - Published: 03 February 2020
   - https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-1929-3
   - Explains the concept of ATAC-seq and discusses the major steps in ATAC-seq analysis, such as quality check and alignment, peak calling, nucleosome position analysis, etc.

3. Supervised classification enables rapid annotation of cell atlases
   - Authors: Hannah A. Pliner, Jay Shendure & Cole Trapnell
   - Published: 09 September 2019
   - https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-1929-3
   - Rapid annotation of cell types in single-cell chromatin accessibility datasets using the tool "Garnett"

4. Transposition of native chromatin for fast and sensitive epigenomic profiling of open chromatin, DNA-binding proteins and nucleosome position
   - Authors: Jason D. Buenrostro, Paul G. Giresi, Lisa C. Zaba, Howard Y. Chang, William J Greenleaf
   - Published: 06 October 2023
   - https://www.nature.com/articles/nmeth.2688
   - ATAC-seq as a method to identify regions of open chromatin, identify nucleosome-bound and nucleosome-free positions in regulatory regions

5. Characterization of the Zebrafish Cell Landscape at Single-Cell Resolution
   - Authors: Mengmeng Jiang, Yanyu Xiao, Weigao E, Lifeng Ma, Jingjing Wang, Haide Chen, Ce Gao, Yuan Liao, Qile Guo, Jinrong Peng, Xiaoping Han, Guoji Guo
   - Published: 01 October 2021
   - https://www.frontiersin.org/articles/10.3389/fcell.2021.743421/full
   - Zebrafish as an important organism to investigate gene regulation and cell lineage specification

6. Central Repository for Management and Application of Marker Gene Lists
   - Author: Micha Frederick Keßler
   - Presented: Dezember 2023
   - Management of the marker lists as a key step in the annotation of the cellular clusters

## Contacts 
For any queries or further discussion, feel free to contact:
- Huijie Li: huijie.li@bioinfsys.uni-giessen.de
- Marta Quintanilla: maria.marta.quintanilla.aguilar@bioinfsys-uni-giessen.de
